File: swift/HikeTimeCore/Sources/HikeTimeCore/Geo.swift
import Foundation

/// Точка (широта, долгота) в градусах.
public struct GeoPoint: Sendable, Codable, Equatable {
    public let lat: Double
    public let lon: Double
    public init(lat: Double, lon: Double) {
        self.lat = lat
        self.lon = lon
    }
}

public enum Geo {

    static let earthR = 6_371_008.8

    public static func haversine(_ a: GeoPoint, _ b: GeoPoint) -> Double {
        let la1 = a.lat * .pi / 180, la2 = b.lat * .pi / 180
        let dLat = (b.lat - a.lat) * .pi / 180
        let dLon = (b.lon - a.lon) * .pi / 180
        let h = sin(dLat / 2) * sin(dLat / 2)
              + cos(la1) * cos(la2) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * earthR * asin(sqrt(h))
    }

    /// Точки через равные интервалы вдоль ломаной: набор высоты не должен
    /// зависеть от того, как густо записан исходный трек.
    public static func resample(_ points: [GeoPoint], step: Double = 25) -> [GeoPoint] {
        guard points.count >= 2 else { return points }
        var out = [points[0]]
        var acc = 0.0
        for i in 1..<points.count {
            var a = points[i - 1]
            let b = points[i]
            var d = haversine(a, b)
            while acc + d >= step {
                let t = (step - acc) / d
                a = GeoPoint(lat: a.lat + (b.lat - a.lat) * t,
                             lon: a.lon + (b.lon - a.lon) * t)
                out.append(a)
                d = haversine(a, b)
                acc = 0
            }
            acc += d
        }
        out.append(points[points.count - 1])
        return out
    }

    /// Скользящее среднее: глушит шум DEM и GPS, иначе набор раздувается.
    public static func smooth(_ values: [Double], window: Int = 5) -> [Double] {
        guard window >= 2, values.count >= window else { return values }
        let half = window / 2
        return values.indices.map { i in
            let lo = max(0, i - half)
            let hi = min(values.count, i + half + 1)
            return values[lo..<hi].reduce(0, +) / Double(hi - lo)
        }
    }

    /// Набор и сброс с порогом: изменения мельче порога — шум.
    public static func gainLoss(_ elevations: [Double], threshold: Double = 2)
        -> (gain: Double, loss: Double) {
        guard var ref = elevations.first else { return (0, 0) }
        var gain = 0.0, loss = 0.0
        for e in elevations.dropFirst() {
            let d = e - ref
            if d > threshold { gain += d; ref = e }
            else if d < -threshold { loss -= d; ref = e }
        }
        return (gain, loss)
    }

    /// Профиль высот -> сегменты для модели времени.
    public static func segments(points: [GeoPoint], elevations: [Double]) -> [Segment] {
        precondition(points.count == elevations.count)
        var out: [Segment] = []
        out.reserveCapacity(max(points.count - 1, 0))
        for i in 1..<max(points.count, 1) {
            out.append(Segment(distance: haversine(points[i - 1], points[i]),
                               dh: elevations[i] - elevations[i - 1]))
        }
        return out
    }
}

File: model/geo.py
"""Геометрия трека: расстояния, ресемплинг, сглаживание профиля."""
import math

R_EARTH = 6371008.8


def haversine(a, b):
    """Расстояние между (lat, lon) в метрах."""
    lat1, lon1 = math.radians(a[0]), math.radians(a[1])
    lat2, lon2 = math.radians(b[0]), math.radians(b[1])
    dlat, dlon = lat2 - lat1, lon2 - lon1
    h = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    return 2 * R_EARTH * math.asin(math.sqrt(h))


def cumulative(points):
    out = [0.0]
    for i in range(1, len(points)):
        out.append(out[-1] + haversine(points[i - 1], points[i]))
    return out


def resample(points, step=25.0):
    """Точки через равные интервалы вдоль ломаной.

    Нужно, чтобы набор высоты не зависел от того, как густо записал GPS.
    """
    if len(points) < 2:
        return list(points)
    cum = cumulative(points)
    total = cum[-1]
    out, target, i = [points[0]], step, 1
    while target < total:
        while i < len(cum) - 1 and cum[i] < target:
            i += 1
        seg = cum[i] - cum[i - 1]
        t = 0.0 if seg == 0 else (target - cum[i - 1]) / seg
        lat = points[i - 1][0] + t * (points[i][0] - points[i - 1][0])
        lon = points[i - 1][1] + t * (points[i][1] - points[i - 1][1])
        out.append((lat, lon))
        target += step
    out.append(points[-1])
    return out


def smooth(values, window=5):
    """Скользящее среднее — глушит шум DEM и GPS, иначе набор высоты раздувается."""
    if window < 2 or len(values) < window:
        return list(values)
    half, out = window // 2, []
    for i in range(len(values)):
        lo, hi = max(0, i - half), min(len(values), i + half + 1)
        out.append(sum(values[lo:hi]) / (hi - lo))
    return out


def gain_loss(elevs, threshold=2.0):
    """Набор и сброс с порогом: изменения мельче порога считаем шумом."""
    gain = loss = 0.0
    ref = elevs[0]
    for e in elevs[1:]:
        d = e - ref
        if d > threshold:
            gain += d
            ref = e
        elif d < -threshold:
            loss += -d
            ref = e
    return gain, loss
